Library/Homebrew/test/cask/system_command_spec.rb

```diff
@@ -1,4 +1,59 @@
 describe Hbc::SystemCommand, :cask do
+  describe "#initialize" do
+    let(:env_args) { ["bash", "-c", 'printf "%s" "${A?}" "${B?}" "${C?}"'] }
+
+    describe "given some environment variables" do
+      subject {
+        described_class.new(
+          "env",
+          args: env_args,
+          env: { "A" => "1", "B" => "2", "C" => "3" },
+          must_succeed: true,
+        )
+      }
+
+      its("run!.stdout") { is_expected.to eq("123") }
+
+      describe "the resulting command line" do
+        it "does not include the given variables" do
+          expect(Open3)
+            .to receive(:popen3)
+            .with(a_hash_including("PATH"), ["env"] * 2, *env_args, {})
+            .and_call_original
+
+          subject.run!
+        end
+      end
+    end
+
+    describe "given some environment variables and sudo: true" do
+      subject {
+        described_class.new(
+          "env",
+          args: env_args,
+          env: { "A" => "1", "B" => "2", "C" => "3" },
+          must_succeed: true,
+          sudo: true,
+        )
+      }
+
+      describe "the resulting command line" do
+        it "includes the given variables explicitly" do
+          expect(Open3)
+            .to receive(:popen3)
+            .with(an_instance_of(Hash), ["/usr/bin/sudo"] * 2,
+                "-E", a_string_starting_with("PATH="),
+                "A=1", "B=2", "C=3", "--", "env", *env_args, {})
+            .and_wrap_original do |original_popen3, *_, &block|
```

Comment:
I think you can add 

```ruby
allow(Open3).to receive(:popen3)
```

before `expect` and in turn leave off `and_wrap_original`.

Reply:
The class under test expects the `#popen3` to return an `IO`-ish object, on which it will call e. g. `#close_write`.
Losing the `and_wrap_original` would cause the test double to return `nil`. The test would then error out.

I’ve been thinking of alternatives, e. g. `and_return` with an `IO` test double; I’m still slightly in favor of `and_wrap_original` for readability.

BTW: I’m probably missing something here but how is a (matching) `allow` before `expect` not a no-op?

Reply:
> BTW: I’m probably missing something here but how is a (matching) allow before expect not a no-op?

No, you're not missing anything. `allow` makes it a no-op, but if the return value is used we cannot do this.